search/nearby_points_sweeper.hpp

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include "base/assert.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/cstdint.hpp"
+#include "std/limits.hpp"
+#include "std/set.hpp"
+#include "std/utility.hpp"
+#include "std/vector.hpp"
+
+namespace search
+{
+// This class can be used to sweep points on a plane that are too
+// close to each other.  Two points are considered to be "too close"
+// when Manhattan distance between them is less than or equal to some
+// preselected epsilon.
+class NearbyPointsSweeper
+{
+public:
+  explicit NearbyPointsSweeper(double eps);
+
+  // Adds a new point (|x|, |y|) on the plane. |index| is used to
+  // identify individual points, and will be reported for survived
+  // points during the Sweep phase.
+  void Add(double x, double y, size_t index);
+
+  // Emits indexes of all survived points. Complexity: O(n * log(n)),
+  // where n is the number of already added points.
+  template <typename TEmitter>
+  void Sweep(TEmitter && emitter)
+  {
+    sort(m_events.begin(), m_events.end());
+
+    set<pair<double, size_t>> line;
+
+    for (auto const & event : m_events)
+    {
+      switch (event.m_type)
+      {
+      case Event::TYPE_SEGMENT_START:
+      {
+        if (line.empty())
+        {
+          line.emplace(event.m_x, event.m_index);
+          break;
+        }
+
+        auto it = line.upper_bound(make_pair(event.m_x, numeric_limits<size_t>::max()));
+
+        bool add = true;
+        while (true)
+        {
+          if (line.empty())
+            break;
+
+          if (it == line.end())
+          {
+            --it;
+            continue;
```

Comment:
I'd prefer to keep it as is. Here's a list of complex conditions, after all of them break or continue happens.